dali/kernels/imgproc/surface.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_SURFACE_H_
+#define DALI_KERNELS_IMGPROC_SURFACE_H_
+
+#include <cuda_runtime.h>
+#include "dali/kernels/tensor_view.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename T>
+struct Surface2D {
+  T *data;
+  int width, height, channels, pixel_stride, row_stride, channel_stride;
+  __host__ __device__ constexpr T &operator()(int x, int y, int c = 0) const {
+    return data[y * row_stride + x * pixel_stride + c * channel_stride];
+  }
+};
+
+template <typename T, typename Storage>
+constexpr Surface2D<T> as_surface_HWC(const TensorView<Storage, T, 3> &t) {
+  return { t.data,
+    static_cast<int>(t.shape[1]), static_cast<int>(t.shape[0]), static_cast<int>(t.shape[2]),
```

Comment:
would prefer to see which member  you are initializing, instead of going up to the definition to find it

Reply:
I agree
```suggestion
    .pixel_stride = static_cast<int>(t.shape[1]), .row_stride = static_cast<int>(t.shape[0]), .channel_stride = static_cast<int>(t.shape[2]),
```
would be more readable here

Reply:
Unfortunately, deisgnated initializers are part of C, not C++. They are added in C++17 or even 20.

Reply:
...and adding constructors will make the type non-POD, which I want to avoid.

Reply:
can we do
```
Surface2D<...> surface;
surface.pixel_stride = ...
surface.row_stride = ...
...
```
or at least break it into lines and put a comment at the end of the line of what is what?

Reply:
non-trivial designated initializers are not supported prior to C++17, but trivial designated initializers seems to be supported (and implemented in both g++ and clang) from 11

Reply:
We can't do what you propose in C++11 (no longer constexpr). I can break lines.

Reply:
@mzient non-trivial designated initializers are not supported prior to C++17, but trivial designated initializers seems to be supported (and implemented in both g++ and clang) from 11